### PR TITLE
Snapshots: Hide the share button

### PR DESCRIPTION
**Pull request by ryantxu**
Dashboard snapshots currently include the share button -- but the results are non-sensical, so we should just hide it.

This avoids creating snapshots, public dashboards, and reports from a snapshot:
before:
image

After:
image

Unlike the other `canXYZ` flags (like canEdit), canShare needs to be explicitly false 🤷🏻 

This PR also refactors the frontend service so we have an explicit entry point for the service, and will more easily allow an additional implementation: https://github.com/grafana/grafana/pull/77667

### CODE REVIEW

File: public/app/features/manage-dashboards/types.ts [hunk 1]
```diff
@@ -2,17 +2,6 @@ import { Dashboard } from '@grafana/schema/src/veneer/dashboard.types';
 
 import { ExternalDashboard } from '../dashboard/components/DashExportModal/DashboardExporter';
 
-export interface Snapshot {
-  created: string;
-  expires: string;
```
Comment (ryantxu) [inline public/app/features/manage-dashboards/types.ts:7]: The values for created/expires/updated are all ignored -- so i removed them from the copy in: https://github.com/grafana/grafana/pull/77763/files#diff-9b1ae42589c4484c3841cc89bedb40095e97ffdc97cd86b05ab9aa1d16b15c4fR5

File: public/app/features/dashboard/services/SnapshotSrv.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  deleteSnapshot: (key: string) => Promise<void>;
+  getSnapshot: (key: string) => Promise<DashboardDTO>;
+}
+
+const legacyDashboardSnapshotSrv: DashboardSnapshotSrv = {
+  getSnapshots: () => backendSrv.get<Snapshot[]>('/api/dashboard/snapshots'),
+  deleteSnapshot: (key: string) => backendSrv.delete('/api/snapshots/' + key),
+  getSnapshot: async (key: string) => {
+    const dto = await backendSrv.get<DashboardDTO>('/api/snapshots/' + key);
+    dto.meta.canShare = false;
```
Comment (ryantxu) [inline public/app/features/dashboard/services/SnapshotSrv.ts:24]: ^^^ the key change

**Review by evictorero — APPROVED** [review]
The refactor looks great. I've tested it and works as expected.

File: public/app/features/dashboard/services/SnapshotSrv.ts (hunk 2 above)
Comment (ryantxu) [reply public/app/features/dashboard/services/SnapshotSrv.ts:33]: ya -- my first version of the PR added it, but that starts the cascade of updating all the swager docs etc

**Comment by marcuscaisey** [comment]
>but the results are non-sensical

I agree that almost all of the options in the share modal don't make sense for snapshots. However, where I work, we use `Export > Save to file` on snapshots when we need to export dashboard data from client environments to investigate production incidents. Would it be possible to add back this functionality please? (Doesn't have to be with the share modal, could just be a button on the navigation bar for snapshots.)

**Pull request merged**